### PR TITLE
sg: add commands to wrap common bazel generating commands

### PR DESCRIPTION
**Pull request by Strum355**
### The Old

Replaces `aspect configure` (used via `bazel configure`) which had a few issues:
- Uses host Go toolchain instead of bazel-managed one
- Didnt use our //:gazelle binary, which is configured with `rules_buf` gazelle rules
- Not able to configure other gazelle rules outside of the fixed set that comes with aspect cli

### The New

`sg bazel configure` and its targets (currently `builds`, `godeps` and `rustdeps`). Passing none only runs `sg bazel configure builds`, the pseudo-target `all` runs all of them.

`sg bazel [other arg]` passes to `bazel` underneath, as a convenience option for if people get more used to running `sg bazel ...` than `bazel` directly (or from shell history)

```
$ bazel-bin/dev/sg/sg_/sg bazel help configure
NAME:
   sg bazel configure - Wrappers around some commands to generate various files required by Bazel

USAGE:
   sg bazel configure [category...]

DESCRIPTION:
   For convenience, a number of Bazel commands are wrapped by this command to update various files required by Bazel.

   Available categories:
     - builds: updates BUILD.bazel files for Go & Typescript targets.
     - godeps: updates the bazel Go dependency targets based on go.mod changes.
     - rustdeps: updates the cargo bazel lockfile.
     - all: catch-all for the above

   If no categories are referenced, then 'builds' is assumed as the default.

OPTIONS:
   --help, -h  show help
```

```
$ bazel-bin/dev/sg/sg_/sg bazel help
Additional commands from sg:
  configure           Wrappers around some commands to generate various files required by Bazel
                                               [bazel release 7.0.0- (@non-git)]
Usage: bazel <command> <options> ...

Available commands:
  analyze-profile     Analyzes build profile data.
  aquery              Analyzes the given targets and queries the action graph.
  build               Builds the specified targets.
  canonicalize-flags  Canonicalizes a list of bazel options.
  clean               Removes output files and optionally stops the server.
  coverage            Generates code coverage report for specified test targets.
  cquery              Loads, analyzes, and queries the specified targets w/ configurations.
  dump                Dumps the internal state of the bazel server process.
  fetch               Fetches external repositories that are prerequisites to the targets.
  help                Prints help for commands, or the index.
  info                Displays runtime info about the bazel server.
  license             Prints the license of this software.
  mobile-install      Installs targets to mobile devices.
  mod                 Queries the Bzlmod external dependency graph
  print_action        Prints the command line args for compiling a file.
  query               Executes a dependency graph query.
  run                 Runs the specified target.
  shutdown            Stops the bazel server.
  sync                Syncs all repositories specified in the workspace file
  test                Builds and runs the specified test targets.
  version             Prints version information for bazel.

Getting more help:
  bazel help <command>
                   Prints help and options for <command>.
  bazel help startup_options
                   Options for the JVM hosting bazel.
  bazel help target-syntax
                   Explains the syntax for specifying targets.
  bazel help info-keys
                   Displays a list of keys used by the info command.
```

## Test plan

`bazel run //dev/sg:sg -- bazel configure` and others

### CODE REVIEW

File: dev/ci/bazel-prechecks.sh [hunk 1]
```diff
@@ -48,7 +48,7 @@ if [[ $EXIT_CODE -ne 0 ]]; then
   BUILD.bazel files need to be updated to match the repository state. You should run the following command and commit the result
 
   ```
-  bazel configure
+  sg bzlgen
```
Comment (Strum355) [inline dev/ci/bazel-prechecks.sh:51]: Itd be nice to use `sg` in this file, using a bootstrapped sg like `bazel run //dev/tools:sg-bootstrap` pulled from https://github.com/sourcegraph/sg/releases, but we risk having some awkwardness when the `sg bzlgen` target changes in a PR but theres no sg release with that update

File: dev/sg/sg_bzlgen.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+		args:   []string{"--only=crate_index"},
+		env:    []string{"CARGO_BAZEL_REPIN=1"},
+		protip: "run with CARGO_BAZEL_ISOLATED=0 for faster (but less sandboxed) repinning.",
+	},
+}
+
+var bzlgenCommand = &cli.Command{
+	Name:      "bzlgen",
+	Usage:     "generates bazel build files that can be generated automatically",
+	UsageText: "sg bzlgen [group...]",
```
Comment (Strum355) [inline dev/sg/sg_bzlgen.go:48]: TODO: description stuff

**Comment by Strum355** [comment]
> So is this a complete replacement for bazel configure in the local flow? Or just CI?

Both perhaps, primarily locally, although in CI we may want to invoke the targets directly instead of through this in case things need to change.

File: dev/sg/sg_bazel.go [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+	- builds: updates BUILD.bazel files for Go & Typescript targets.
+	- godeps: updates the bazel Go dependency targets based on go.mod changes.
+	- rustdeps: updates the cargo bazel lockfile.
+	- all: catch-all for the above
+
+If no categories are referenced, then 'builds' is assumed as the default.`,
+			Before: func(ctx *cli.Context) error {
+				for _, arg := range ctx.Args().Slice() {
+					if _, ok := bzlgenTargets[arg]; !ok && arg != "all" {
+						cli.HandleExitCoder(errors.Errorf("category doesn't exist %q, run `sg bazel configure -help` for full info.", arg))
```
Comment (burmudar) [inline dev/sg/sg_bazel.go:85]:
```suggestion
						cli.HandleExitCoder(errors.Errorf("category doesn't exist %q, run `sg bazel configure --help` for full info.", arg))
```

File: dev/ci/bazel-prechecks.sh [hunk 4]
```diff
@@ -30,16 +30,16 @@ function generate_diff_artifact() {
 
 trap generate_diff_artifact EXIT
 
-echo "--- :bazel: Running bazel configure"
-bazel "${bazelrc[@]}" configure
+echo "--- :bazel: Running bazel run //:gazelle"
+bazel "${bazelrc[@]}" run //:gazelle
```
Comment (burmudar) [inline dev/ci/bazel-prechecks.sh:34]: Nice! This will make my juggling in https://github.com/sourcegraph/sourcegraph/pull/59968/files pointless 🚀

**Review by burmudar — APPROVED** [review]
Nice! I like the egonomics this brings!

**Pull request merged**